src/main/scala/com/mozilla/telemetry/learning/federated/FederatedLearningSearchOptimizerS3Sink.scala

```diff
@@ -0,0 +1,151 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+package com.mozilla.telemetry.learning.federated
+
+import com.mozilla.telemetry.learning.federated.FederatedLearningSearchOptimizerConstants.{NumberOfFeatures, StartingLearningRate, StartingWeights}
+import com.mozilla.telemetry.streaming.FrecencyUpdateAggregate
+import org.apache.commons.io.IOUtils
+import org.apache.hadoop.fs.Path
+import org.apache.spark.deploy.SparkHadoopUtil
+import org.apache.spark.sql.execution.streaming.Sink
+import org.apache.spark.sql.sources.StreamSinkProvider
+import org.apache.spark.sql.streaming.OutputMode
+import org.apache.spark.sql.{DataFrame, SQLContext}
+import org.json4s.DefaultFormats
+import org.json4s.jackson.Serialization
+
+object FederatedLearningSearchOptimizerConstants {
+  // https://github.com/florian/federated-learning-addon/blob/master/addon/prefs.js#L26
+  val StartingWeights: Array[Double] = Array(4, 14, 31, 90, 100, 70, 50, 30, 10, 0, 0, 100, 2000, 75, 0, 0, 0, 25, 0, 140, 200, 0)
+  val NumberOfFeatures: Int = StartingWeights.length
+  val StartingLearningRate: Int = 2
+}
+
+class FederatedLearningSearchOptimizerS3Sink(outputPath: String, stateCheckpointPath: String) extends Sink {
+
+  private[federated] var state: OptimisationState = initState()
+
+  override def addBatch(batchId: Long, data: DataFrame): Unit = {
+    val aggregates = data.collect().map(FrecencyUpdateAggregate(_))
+
+    val iteration: Long = state.iteration
+
+    val ord = Ordering.by((_: FrecencyUpdateAggregate).modelVersion)
+    aggregates.filter(_.studyVariation == "treatment").filter(_.modelVersion.toLong >= iteration).reduceOption(ord.min) match {
```

Comment:
Isn't this, pragmatically, equivalent to filtering by `_.modelVersion == iteration`? I'm also curious why you chose `>=` -- when would you expect to see a model version that's greater than the current iteration?

Reply:
Good question - I don't know when this might occur (most likely when we misconfigure something).
I think I wanted to be more careful by not dropping too much data, but when I think of it now I'm not sure if it makes much difference.
I'll think about it, let me know if you think it better to switch to equality comparison here.